src/groupeddataframe/fastaggregates.jl

```diff
@@ -163,18 +163,30 @@ function groupreduce!(res::AbstractVector, f, op, condf, adjust, checkempty::Boo
         counts = zeros(Int, n)
     end
     groups = gd.groups
-    @inbounds for i in eachindex(incol, groups)
-        gix = groups[i]
-        x = incol[i]
-        if gix > 0 && (condf === nothing || condf(x))
-            # this check should be optimized out if U is not Any
-            if eltype(res) === Any && !isassigned(res, gix)
-                res[gix] = f(x, gix)
-            else
-                res[gix] = op(res[gix], f(x, gix))
-            end
-            if adjust !== nothing || checkempty
-                counts[gix] += 1
+    if VERSION >= v"1.4" && Threads.nthreads() > 1
+        batchsize = 100_000
+    else
+        batchsize = typemax(Int)
+    end
+    for batch in Iterators.partition(eachindex(incol, groups), batchsize)
+        # Allow other tasks to do garbage collection while this one runs
+        @static if VERSION >= v"1.4"
+            GC.safepoint()
```

Comment:
The only issue is that `GC.safepoint` requires that the thread in which we call it is non-allocating. Are we sure this is the case here? (maybe we should ask on Slack - as I do not know the internals of how `GC.safepoint` works, but it seems that if the thread is allocating it would be not safe to add it)

Reply:
I don't think that's right. The docstring says "Inserts a point in the program where garbage collection may run. [...] Calling this function periodically in non-allocating threads allows garbage collection to run." It's just that it's redundant if the thread allocates, as the GC can already run automatically at these points if needed.

Reply:
OK - probably you are right. I will test it then again tomorrow.